### ISSUE TITLE
Using data augmentation on YOLOv8-pose

### ISSUE THREAD

**Issue by dayong233** [ultralytics/ultralytics#12646]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I currently have 100 real images and 2000 rendered images with corresponding labels. 
I trained those 100 real pictures individually to get train1's best.pt to predict the real pictures. It recognizes Bboxes fairly well, but keypoints has poor accuracy, which is to be expected. 
Then I trained the 2000 rendered pictures separately and got train2's best.pt to predict the real pictures. Sometimes I could not identify bbox well, but the accuracy of keypoints I identified was quite ideal.
Now I want to train them together, but since the number of real pictures is so small, I want to know how to enhance them with data to achieve better training results.

**Comment by dayong233** [comment]
Before put forward this question I've ever seen a similar problem: 
[](https://github.com/ultralytics/ultralytics/issues/2412)
I also found the parts to implement data enhancement:
[](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/data/augment.py)

But I don't know how to do data enhancement on part of my data set. Do I need to write my own data augmentation script if necessary?

**Comment by glenn-jocher** [comment]
Hello! It's great to see that you've done your research on data augmentation. 👍

You can indeed utilize our existing data augmentation functionalities within YOLOv8, which are defined in `augment.py`. There's no need to write an entire script from scratch. You can adjust the pipeline to emphasize specific augmentations more suited to your smaller dataset of real images.

You can modify the data augmentation settings directly in your dataset YAML file or apply them selectively by modifying the augmentation conditions in the `augment.py`. For example, you could introduce conditions to apply stronger or different augmentations based on the image source.

Here's a simple example of how you might conditionally apply augmentations within your augmentation function:

```python
import numpy as np

def custom_augmentation(image, label, source):
    if 'real' in source:
        # Apply specific augmentations for real images
        image = np.fliplr(image)  # Example augmentation
    else:
        # Standard augmentation for rendered images
        image = np.flipud(image)  # Another example augmentation
    return image, label
```

You would then use this function during your dataset loading or preprocessing phase, ensuring that it is part of your training pipeline.

Good luck with enhancing your dataset!

**Comment by medphisiker** [comment]
Hello, Glen )

I have the same question. Thank you for your response and explanations.
However, I am still unsure about how to proceed.

> You can modify the data augmentation settings directly in your dataset YAML file.

The documentation([link](https://docs.ultralytics.com/usage/cfg/#augmentation-settings)) also indicates that we can specify augmentations directly in the YAML file of our dataset.

But unfortunately I did not find a description of how to do this in the YAML dataset file.
I looked at examples of dataset descriptions from docs([link](https://docs.ultralytics.com/datasets/)) and dataset configs([link](https://github.com/ultralytics/ultralytics/tree/main/ultralytics/cfg/datasets)).
But there are clean datasets and there was no example of how to describe augmentation.
I also looked at the documentation describing the general structure of the Pose Estimation Datasets file ([link](https://docs.ultralytics.com/datasets/pose/#ultralytics-yolo-format)).

I'm sure the documentation has a description of the data.yaml file structure with all possible parameters, but I can't find it.

**Comment by medphisiker** [comment]
Update

I found the file (`https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/default.yaml`)

Then, according to this file, I can add it to my data.yaml string
```
train: train  # train images (relative to 'path') 210 images
val: val  # val images (relative to 'path') 53 images

# Keypoints
kpt_shape: [3, 3]  # number of keypoints, number of dims (2 for x,y or 3 for x,y,visible)

# Classes
names:
  0: class1
  1: class2
  2: class13
  
# Hyperparameters
flipud: 0.0 # (float) image flip up-down (probability)
```

In order to apply vertical flip augmentation ? =)

**Comment by medphisiker** [comment]
**UPDATE**

I found a manual in the form of a Markdown file about configuration files([link](https://github.com/ultralytics/ultralytics/blob/c8b6a8bc37ce196a6b7e0c153539402db4383c16/docs/en/usage/cfg.md)).

    ```python
    from ultralytics import YOLO

    # Load a YOLOv8 model from a pre-trained weights file
    model = YOLO("yolov8n.pt")

    # Run MODE mode using the custom arguments ARGS (guess TASK)
    model.MODE(ARGS)
    ```
    
* ARGS (optional) are arg=value pairs like imgsz=640 that override defaults.
Default ARG values are defined on this page from the cfg/defaults.yaml [file](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/default.yaml).

So I can make it even easier? =)

```
from ultralytics import YOLO

model = YOLO("yolov8l-pose.pt")

model.train(
    data="path/to/data.yaml",
    flipud=0.5
)
```
or any other option for augmentation specified in the documentation? ([link](https://docs.ultralytics.com/usage/cfg/#augmentation-settings)) =)

**Comment by glenn-jocher** [comment]
Hello! 😊

Yes, you're absolutely right! You can directly pass augmentation parameters like `flipud=0.5` when calling the `train()` method, just as you've shown in your example. This will override the default settings for that particular training session.

```python
from ultralytics import YOLO

model = YOLO("yolov8l-pose.pt")

model.train(
    data="path/to/data.yaml",
    flipud=0.5  # Apply vertical flip augmentation with 50% probability
)
```

This approach is quite flexible and allows you to experiment with different augmentation settings easily. Just make sure that the parameters you pass are supported as per the [augmentation settings documentation](https://docs.ultralytics.com/usage/cfg/#augmentation-settings).

Keep up the great work, and happy training! 🚀

**Comment by medphisiker** [comment]
Thank you for the clarification, Glenn)
This approach is really more flexible and convenient! 🚀

**Comment by glenn-jocher** [comment]
@medphisiker you're welcome! I'm glad you found the approach flexible and convenient. If you have any more questions or need further assistance as you continue with your project, feel free to reach out. Happy coding! 🌟
